### PR TITLE
Add toJSON method

### PR DESCRIPTION
**Pull request by cemremengu**
I have been been meaning to send a PR for this for sometime. This is an initial implementation (it is missing `plugins`)

Just wanted to ask whether this is something that is acceptable for `toJSON` method or are there any extras or it should be a plugin etc. etc.. If this is OK, I will continue to work on it.

### CODE REVIEW

**Comment by StarpTech** [comment]
I don't get it how is related to a `toJSON` function? Could you explain it?

**Comment by cemremengu** [comment]
Yes it doesn't look like `toJSON` at all now 😄 As far as I understand, `fastify.toJSON` needs to return a json representation of the internal state/config of the fastify object such as routes, plugins, middleware and hooks. This PR just includes the approach for collecting that data. You can assume that `fastify.toJSON` will return the `fastify.state` object.

I am curious whether this is a step in the right direction for building the json object to be returned.

File: fastify.js [hunk 1]
```diff
@@ -76,7 +76,12 @@ function build (options) {
   }
 
   const fastify = {
-    [kChildren]: []
+    [kChildren]: [],
+    state: {
+      hooks: new Hooks(),
+      middlewares: [],
```
Comment (mcollina) [inline fastify.js]: middlewares are structured in a tree, i.e. they are plugin/prefix specific

File: test/toJSON.test.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        },
+        hooks: {
+          onRequest: ['onRequestHook'],
+          preHandler: ['preHandlerHook'],
+          onResponse: ['onResponseHook'],
+          onSend: ['onSendHook'],
+          preValidation: [],
+          onError: []
+        },
+        plugins: ['cemre'] }))
```
Comment (cemremengu) [inline test/toJSON.test.js]: @mcollina any idea why this prints my name? 😄

File: lib/toJSON.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    this[kHooks][key].forEach(h => {
+      const hookName = h.name.split(' ')[1]
+      if (hookName !== '') {
+        hooks[key].push(hookName)
+      }
+    })
+  })
+
+  this[pluginUtils.registeredPlugins].forEach(name => {
+    plugins.push(name)
```
Comment (cemremengu) [inline lib/toJSON.js]: @mcollina any idea why this returns my name?  😄

**Comment by mcollina** [comment]
I think the output follows the wrong structure, this is something more along the lines of what I was thinking

```js
{
  plugins: [{
    name: 'abcde',
    prefix: '',
    encapsulated: true,
    middlewares: [
      'helloWorld',
      'routeABC'
    ]
        },
        hooks: {
          onRequest: ['onRequestHook'],
          preHandler: ['preHandlerHook'],
          onResponse: ['onResponseHook'],
          onSend: ['onSendHook'],
          preValidation: [],
          onError: []
        },
        plugins: [
   // nested as will
  ],
        routes: {
          '/v1/hello/:world': {
            'post': {
              'method': 'POST',
              'schema': {},
              'url': '/v1/hello/:world',
              'logLevel': 'info',
              'prefix': '/v1',
              'bodyLimit': 2000
            }
          }
        }
      }
```

**Comment by delvedor** [comment]
I think we should close this pr (the code is too much outdated) and open an issue to discuss the feature.

**Pull request closed without merging**